include/range/v3/view_adaptor.hpp

```diff
@@ -316,24 +331,29 @@ namespace ranges
                 // the underlying iterator.
                 BaseSent base() const
                 {
-                    return this->get().first;
+                    return this->get().first();
                 }
             };
 
             using compressed_pair<BaseSent, Adapt>::first;
             using compressed_pair<BaseSent, Adapt>::second;
 
-            template<typename I, typename IA, typename A = Adapt,
-                typename R = decltype(std::declval<A>().empty(std::declval<I>(), std::declval<IA>(), first))>
+            template<typename I, typename IA, typename A = Adapt, typename = decltype(
+                std::declval<A const &>().empty(
+                    std::declval<I const &>(),
+                    std::declval<IA const &>(),
+                    std::declval<BaseSent const &>()))>
             constexpr bool equal_(adaptor_cursor<I, IA> const &that, int) const
             {
-                return second.empty(that.first, that.second, first);
+                return second().empty(that.first(), that.second(), first());
             }
-            template<typename I, typename IA, typename A = Adapt,
-                typename R = decltype(std::declval<A>().empty(std::declval<I>(), first))>
+            template<typename I, typename IA, typename A = Adapt, typename = decltype(
+                std::declval<A const &>().empty(
+                    std::declval<I const &>(),
+                    std::declval<BaseSent const &>()))>
             constexpr bool equal_(adaptor_cursor<I, IA> const &that, long) const
             {
-                return second.empty(that.first, first);
+                return second().empty(that.first(), first());
```

Comment:
I remember being nauseous when I wrote this code and having every intention to come back to it and make it suck less. Now it sucks more (which is the nature of sucky code under maintenance). I welcome any thoughts you might have. `view_adaptor` in general has always been a problematic design issue.